File: php/rsq_sdk.php
<?php
declare(strict_types=1);

// Rsq SDK

require_once __DIR__ . '/utility/struct/Struct.php';
require_once __DIR__ . '/core/UtilityType.php';
require_once __DIR__ . '/core/Spec.php';
require_once __DIR__ . '/core/Helpers.php';

// Load utility registration
require_once __DIR__ . '/utility/Register.php';

// Load config and features
require_once __DIR__ . '/config.php';
require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/features.php';

use Voxgig\Struct\Struct;

class RsqSDK
{
    public string $mode;
    public array $features;
    public ?array $options;

    private $_utility;
    private $_rootctx;

    public function __construct(array $options = [])
    {
        $this->mode = "live";
        $this->features = [];
        $this->options = null;

        $utility = new RsqUtility();
        $this->_utility = $utility;

        $config = RsqConfig::make_config();

        $this->_rootctx = ($utility->make_context)([
            "client" => $this,
            "utility" => $utility,
            "config" => $config,
            "options" => $options ?? [],
            "shared" => [],
        ], null);

        $this->options = ($utility->make_options)($this->_rootctx);

        if (Struct::getpath($this->options, "feature.test.active") === true) {
            $this->mode = "test";
        }

        $this->_rootctx->options = $this->options;

        // Add features from config.
        $feature_opts = RsqHelpers::to_map(Struct::getprop($this->options, "feature"));
        if ($feature_opts) {
            $items = Struct::items($feature_opts);
            if ($items) {
                foreach ($items as $item) {
                    $fname = $item[0];
                    $fopts = RsqHelpers::to_map($item[1]);
                    if ($fopts && isset($fopts["active"]) && $fopts["active"] === true) {
                        ($utility->feature_add)($this->_rootctx, RsqFeatures::make_feature($fname));
                    }
                }
            }
        }

        // Add extension features.
        $extend_val = Struct::getprop($this->options, "extend");
        if (is_array($extend_val)) {
            foreach ($extend_val as $f) {
                if (is_object($f) && method_exists($f, 'get_name')) {
                    ($utility->feature_add)($this->_rootctx, $f);
                }
            }
        }

        // Initialize features.
        foreach ($this->features as $f) {
            ($utility->feature_init)($this->_rootctx, $f);
        }

        ($utility->feature_hook)($this->_rootctx, "PostConstruct");
    }

    public function options_map(): array
    {
        $out = Struct::clone($this->options);
        return is_array($out) ? $out : [];
    }

    public function get_utility()
    {
        return RsqUtility::copy($this->_utility);
    }

    public function get_root_ctx()
    {
        return $this->_rootctx;
    }

    public function prepare(array $fetchargs = []): mixed
    {
        $utility = $this->_utility;
        $fetchargs = $fetchargs ?? [];

        $ctrl = RsqHelpers::to_map(Struct::getprop($fetchargs, "ctrl")) ?? [];

        $ctx = ($utility->make_context)([
            "opname" => "prepare",
            "ctrl" => $ctrl,
        ], $this->_rootctx);

        $opts = $this->options;
        $path = Struct::getprop($fetchargs, "path") ?? "";
        $path = is_string($path) ? $path : "";
        $method_val = Struct::getprop($fetchargs, "method") ?? "GET";
        $method_val = is_string($method_val) ? $method_val : "GET";
        $params = RsqHelpers::to_map(Struct::getprop($fetchargs, "params")) ?? [];
        $query = RsqHelpers::to_map(Struct::getprop($fetchargs, "query")) ?? [];
        $headers = ($utility->prepare_headers)($ctx);

        $base = Struct::getprop($opts, "base") ?? "";
        $base = is_string($base) ? $base : "";
        $prefix = Struct::getprop($opts, "prefix") ?? "";
        $prefix = is_string($prefix) ? $prefix : "";
        $suffix = Struct::getprop($opts, "suffix") ?? "";
        $suffix = is_string($suffix) ? $suffix : "";

        $ctx->spec = new RsqSpec([
            "base" => $base, "prefix" => $prefix, "suffix" => $suffix,
            "path" => $path, "method" => $method_val,
            "params" => $params, "query" => $query, "headers" => $headers,
            "body" => Struct::getprop($fetchargs, "body"),
            "step" => "start",
        ]);

        // Merge user-provided headers.
        $uh = Struct::getprop($fetchargs, "headers");
        if (is_array($uh)) {
            foreach ($uh as $k => $v) {
                $ctx->spec->headers[$k] = $v;
            }
        }

        [$_, $err] = ($utility->prepare_auth)($ctx);
        if ($err) {
            return ($utility->make_error)($ctx, $err);
        }

        [$fetchdef, $fd_err] = ($utility->make_fetch_def)($ctx);
        if ($fd_err) {
            return ($utility->make_error)($ctx, $fd_err);
        }
        return $fetchdef;
    }

    public function direct(array $fetchargs = []): mixed
    {
        $utility = $this->_utility;

        // direct() is the raw-HTTP escape hatch: it never throws, it returns
        // an {ok, err, ...} dict. prepare() now raises on error, so catch it
        // and surface the failure through the dict instead.
        try {
            $fetchdef = $this->prepare($fetchargs);
        } catch (\Throwable $err) {
            return ["ok" => false, "err" => $err];
        }

        $fetchargs = $fetchargs ?? [];
        $ctrl = RsqHelpers::to_map(Struct::getprop($fetchargs, "ctrl")) ?? [];

        $ctx = ($utility->make_context)([
            "opname" => "direct",
            "ctrl" => $ctrl,
        ], $this->_rootctx);

        $url = $fetchdef["url"] ?? "";
        [$fetched, $fetch_err] = ($utility->fetcher)($ctx, $url, $fetchdef);

        if ($fetch_err) {
            return ["ok" => false, "err" => $fetch_err];
        }

        if ($fetched === null) {
            return [
                "ok" => false,
                "err" => $ctx->make_error("direct_no_response", "response: undefined"),
            ];
        }

        if (is_array($fetched)) {
            $status = RsqHelpers::to_int(Struct::getprop($fetched, "status"));
            $headers = Struct::getprop($fetched, "headers") ?? [];

            // No-body responses (204, 304) and explicit zero content-length
            // must skip JSON parsing — calling json() on an empty body errors.
            $content_length = is_array($headers) ? ($headers["content-length"] ?? null) : null;
            $no_body = $status === 204 || $status === 304 || (string)$content_length === "0";

            $json_data = null;
            if (!$no_body) {
                $jf = Struct::getprop($fetched, "json");
                if (is_callable($jf)) {
                    try {
                        $json_data = $jf();
                    } catch (\Throwable $e) {
                        // Non-JSON body — leave data null but keep status/ok.
                        $json_data = null;
                    }
                }
            }

            return [
                "ok" => $status >= 200 && $status < 300,
                "status" => $status,
                "headers" => Struct::getprop($fetched, "headers"),
                "data" => $json_data,
            ];
        }

        return [
            "ok" => false,
            "err" => $ctx->make_error("direct_invalid", "invalid response type"),
        ];
    }


    private $_category = null;

    // Canonical facade: $client->Category()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->category()
    // resolves here too.
    public function Category($data = null)
    {
        require_once __DIR__ . '/entity/category_entity.php';
        if ($data === null) {
            if ($this->_category === null) {
                $this->_category = new CategoryEntity($this, null);
            }
            return $this->_category;
        }
        return new CategoryEntity($this, $data);
    }


    private $_country_of_asylum = null;

    // Canonical facade: $client->CountryOfAsylum()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->country_of_asylum()
    // resolves here too.
    public function CountryOfAsylum($data = null)
    {
        require_once __DIR__ . '/entity/country_of_asylum_entity.php';
        if ($data === null) {
            if ($this->_country_of_asylum === null) {
                $this->_country_of_asylum = new CountryOfAsylumEntity($this, null);
            }
            return $this->_country_of_asylum;
        }
        return new CountryOfAsylumEntity($this, $data);
    }


    private $_country_of_origin = null;

    // Canonical facade: $client->CountryOfOrigin()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->country_of_origin()
    // resolves here too.
    public function CountryOfOrigin($data = null)
    {
        require_once __DIR__ . '/entity/country_of_origin_entity.php';
        if ($data === null) {
            if ($this->_country_of_origin === null) {
                $this->_country_of_origin = new CountryOfOriginEntity($this, null);
            }
            return $this->_country_of_origin;
        }
        return new CountryOfOriginEntity($this, $data);
    }


    private $_country_of_resettlement = null;

    // Canonical facade: $client->CountryOfResettlement()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->country_of_resettlement()
    // resolves here too.
    public function CountryOfResettlement($data = null)
    {
        require_once __DIR__ . '/entity/country_of_resettlement_entity.php';
        if ($data === null) {
            if ($this->_country_of_resettlement === null) {
                $this->_country_of_resettlement = new CountryOfResettlementEntity($this, null);
            }
            return $this->_country_of_resettlement;
        }
        return new CountryOfResettlementEntity($this, $data);
    }


    private $_demographic = null;

    // Canonical facade: $client->Demographic()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->demographic()
    // resolves here too.
    public function Demographic($data = null)
    {
        require_once __DIR__ . '/entity/demographic_entity.php';
        if ($data === null) {
            if ($this->_demographic === null) {
                $this->_demographic = new DemographicEntity($this, null);
            }
            return $this->_demographic;
        }
        return new DemographicEntity($this, $data);
    }


    private $_departure = null;

    // Canonical facade: $client->Departure()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->departure()
    // resolves here too.
    public function Departure($data = null)
    {
        require_once __DIR__ . '/entity/departure_entity.php';
        if ($data === null) {
            if ($this->_departure === null) {
                $this->_departure = new DepartureEntity($this, null);
            }
            return $this->_departure;
        }
        return new DepartureEntity($this, $data);
    }


    private $_helper = null;

    // Canonical facade: $client->Helper()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->helper()
    // resolves here too.
    public function Helper($data = null)
    {
        require_once __DIR__ . '/entity/helper_entity.php';
        if ($data === null) {
            if ($this->_helper === null) {
                $this->_helper = new HelperEntity($this, null);
            }
            return $this->_helper;
        }
        return new HelperEntity($this, $data);
    }


    private $_region = null;

    // Canonical facade: $client->Region()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->region()
    // resolves here too.
    public function Region($data = null)
    {
        require_once __DIR__ . '/entity/region_entity.php';
        if ($data === null) {
            if ($this->_region === null) {
                $this->_region = new RegionEntity($this, null);
            }
            return $this->_region;
        }
        return new RegionEntity($this, $data);
    }


    private $_submission = null;

    // Canonical facade: $client->Submission()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->submission()
    // resolves here too.
    public function Submission($data = null)
    {
        require_once __DIR__ . '/entity/submission_entity.php';
        if ($data === null) {
            if ($this->_submission === null) {
                $this->_submission = new SubmissionEntity($this, null);
            }
            return $this->_submission;
        }
        return new SubmissionEntity($this, $data);
    }


    private $_url_fetch = null;

    // Canonical facade: $client->UrlFetch()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->url_fetch()
    // resolves here too.
    public function UrlFetch($data = null)
    {
        require_once __DIR__ . '/entity/url_fetch_entity.php';
        if ($data === null) {
            if ($this->_url_fetch === null) {
                $this->_url_fetch = new UrlFetchEntity($this, null);
            }
            return $this->_url_fetch;
        }
        return new UrlFetchEntity($this, $data);
    }


    private $_year = null;

    // Canonical facade: $client->Year()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->year()
    // resolves here too.
    public function Year($data = null)
    {
        require_once __DIR__ . '/entity/year_entity.php';
        if ($data === null) {
            if ($this->_year === null) {
                $this->_year = new YearEntity($this, null);
            }
            return $this->_year;
        }
        return new YearEntity($this, $data);
    }



    public static function test(?array $testopts = null, ?array $sdkopts = null): self
    {
        $sdkopts = $sdkopts ?? [];
        $sdkopts = Struct::clone($sdkopts);
        $sdkopts = is_array($sdkopts) ? $sdkopts : [];

        $testopts = $testopts ?? [];
        $testopts = Struct::clone($testopts);
        $testopts = is_array($testopts) ? $testopts : [];
        $testopts["active"] = true;

        if (!isset($sdkopts["feature"])) {
            $sdkopts["feature"] = [];
        }
        $sdkopts["feature"]["test"] = $testopts;

        $sdk = new RsqSDK($sdkopts);
        $sdk->mode = "test";
        return $sdk;
    }
}

File: lua/rsq_sdk.lua
-- Rsq SDK

local vs = require("utility.struct.struct")
local Utility = require("core.utility_type")
local Spec = require("core.spec")
local helpers = require("core.helpers")

-- Load utility registration (populates Utility._registrar)
require("utility.register")

-- Load features
local BaseFeature = require("feature.base_feature")
local features_factory = require("features")


local RsqSDK = {}
RsqSDK.__index = RsqSDK


local function _make_feature(name)
  local factory = features_factory[name]
  if factory ~= nil then
    return factory()
  end
  return features_factory.base()
end

RsqSDK._make_feature = _make_feature


function RsqSDK.new(options)
  local self = setmetatable({}, RsqSDK)
  self.mode = "live"
  self.features = {}
  self.options = nil

  local utility = Utility.new()
  self._utility = utility

  local config = require("config")()

  self._rootctx = utility.make_context({
    client = self,
    utility = utility,
    config = config,
    options = options or {},
    shared = {},
  }, nil)

  self.options = utility.make_options(self._rootctx)

  if vs.getpath(self.options, "feature.test.active") == true then
    self.mode = "test"
  end

  self._rootctx.options = self.options

  -- Add features in the resolved order (make_options puts an explicit list
  -- order first, else defaults to test-first). Ordering matters: the `test`
  -- feature installs the base mock transport and the transport features
  -- (retry/cache/netsim/proxy/ratelimit) wrap whatever is current, so `test`
  -- must be added before them to sit at the base of the chain.
  local feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
  if feature_opts ~= nil then
    local featureorder = vs.getpath(self.options, "__derived__.featureorder")
    if type(featureorder) == "table" then
      for _, fname in ipairs(featureorder) do
        local fopts = helpers.to_map(feature_opts[fname])
        if fopts ~= nil and fopts["active"] == true then
          utility.feature_add(self._rootctx, _make_feature(fname))
        end
      end
    end
  end

  -- Add extension features.
  local extend = vs.getprop(self.options, "extend")
  if type(extend) == "table" then
    for _, f in ipairs(extend) do
      if type(f) == "table" and type(f.get_name) == "function" then
        utility.feature_add(self._rootctx, f)
      end
    end
  end

  -- Initialize features.
  for _, f in ipairs(self.features) do
    utility.feature_init(self._rootctx, f)
  end

  utility.feature_hook(self._rootctx, "PostConstruct")

  -- #BuildFeatures

  return self
end


function RsqSDK:options_map()
  local out = vs.clone(self.options)
  if type(out) == "table" then
    return out
  end
  return {}
end


function RsqSDK:get_utility()
  return Utility.copy(self._utility)
end


function RsqSDK:get_root_ctx()
  return self._rootctx
end


function RsqSDK:prepare(fetchargs)
  local utility = self._utility

  fetchargs = fetchargs or {}

  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "prepare",
    ctrl = ctrl,
  }, self._rootctx)

  local options = self.options

  local path = vs.getprop(fetchargs, "path") or ""
  if type(path) ~= "string" then path = "" end

  local method = vs.getprop(fetchargs, "method") or "GET"
  if type(method) ~= "string" then method = "GET" end

  local params = helpers.to_map(vs.getprop(fetchargs, "params")) or {}
  local query = helpers.to_map(vs.getprop(fetchargs, "query")) or {}

  local headers = utility.prepare_headers(ctx)

  local base = vs.getprop(options, "base") or ""
  if type(base) ~= "string" then base = "" end
  local prefix = vs.getprop(options, "prefix") or ""
  if type(prefix) ~= "string" then prefix = "" end
  local suffix = vs.getprop(options, "suffix") or ""
  if type(suffix) ~= "string" then suffix = "" end

  ctx.spec = Spec.new({
    base = base,
    prefix = prefix,
    suffix = suffix,
    path = path,
    method = method,
    params = params,
    query = query,
    headers = headers,
    body = vs.getprop(fetchargs, "body"),
    step = "start",
  })

  -- Merge user-provided headers.
  local uh = vs.getprop(fetchargs, "headers")
  if type(uh) == "table" then
    for k, v in pairs(uh) do
      ctx.spec.headers[k] = v
    end
  end

  local _, err = utility.prepare_auth(ctx)
  if err ~= nil then
    return nil, err
  end

  return utility.make_fetch_def(ctx)
end


function RsqSDK:direct(fetchargs)
  local utility = self._utility

  local fetchdef, err = self:prepare(fetchargs)
  if err ~= nil then
    return { ok = false, err = err }, nil
  end

  fetchargs = fetchargs or {}
  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "direct",
    ctrl = ctrl,
  }, self._rootctx)

  local url = fetchdef["url"] or ""
  local fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

  if fetch_err ~= nil then
    return { ok = false, err = fetch_err }, nil
  end

  if fetched == nil then
    return {
      ok = false,
      err = ctx:make_error("direct_no_response", "response: undefined"),
    }, nil
  end

  if type(fetched) == "table" then
    local status = helpers.to_int(vs.getprop(fetched, "status"))
    local headers = vs.getprop(fetched, "headers") or {}

    -- No-body responses (204, 304) and explicit zero content-length
    -- must skip JSON parsing — calling json() on an empty body errors.
    local content_length = nil
    if type(headers) == "table" then
      content_length = headers["content-length"]
    end
    local no_body = status == 204 or status == 304 or tostring(content_length) == "0"

    local json_data = nil
    if not no_body then
      local jf = vs.getprop(fetched, "json")
      if type(jf) == "function" then
        local ok, result = pcall(jf)
        if ok then
          json_data = result
        end
        -- Non-JSON body: json_data stays nil, status/headers preserved.
      end
    end

    return {
      ok = status >= 200 and status < 300,
      status = status,
      headers = headers,
      data = json_data,
    }, nil
  end

  return {
    ok = false,
    err = ctx:make_error("direct_invalid", "invalid response type"),
  }, nil
end



-- Idiomatic facade: client:Category():list() / client:Category():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function RsqSDK:Category(data)
  local EntityMod = require("entity.category_entity")
  if data == nil then
    if self._category == nil then
      self._category = EntityMod.new(self, nil)
    end
    return self._category
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:CountryOfAsylum():list() / client:CountryOfAsylum():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function RsqSDK:CountryOfAsylum(data)
  local EntityMod = require("entity.country_of_asylum_entity")
  if data == nil then
    if self._country_of_asylum == nil then
      self._country_of_asylum = EntityMod.new(self, nil)
    end
    return self._country_of_asylum
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:CountryOfOrigin():list() / client:CountryOfOrigin():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function RsqSDK:CountryOfOrigin(data)
  local EntityMod = require("entity.country_of_origin_entity")
  if data == nil then
    if self._country_of_origin == nil then
      self._country_of_origin = EntityMod.new(self, nil)
    end
    return self._country_of_origin
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:CountryOfResettlement():list() / client:CountryOfResettlement():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function RsqSDK:CountryOfResettlement(data)
  local EntityMod = require("entity.country_of_resettlement_entity")
  if data == nil then
    if self._country_of_resettlement == nil then
      self._country_of_resettlement = EntityMod.new(self, nil)
    end
    return self._country_of_resettlement
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Demographic():list() / client:Demographic():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function RsqSDK:Demographic(data)
  local EntityMod = require("entity.demographic_entity")
  if data == nil then
    if self._demographic == nil then
      self._demographic = EntityMod.new(self, nil)
    end
    return self._demographic
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Departure():list() / client:Departure():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function RsqSDK:Departure(data)
  local EntityMod = require("entity.departure_entity")
  if data == nil then
    if self._departure == nil then
      self._departure = EntityMod.new(self, nil)
    end
    return self._departure
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Helper():list() / client:Helper():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function RsqSDK:Helper(data)
  local EntityMod = require("entity.helper_entity")
  if data == nil then
    if self._helper == nil then
      self._helper = EntityMod.new(self, nil)
    end
    return self._helper
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Region():list() / client:Region():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function RsqSDK:Region(data)
  local EntityMod = require("entity.region_entity")
  if data == nil then
    if self._region == nil then
      self._region = EntityMod.new(self, nil)
    end
    return self._region
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Submission():list() / client:Submission():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function RsqSDK:Submission(data)
  local EntityMod = require("entity.submission_entity")
  if data == nil then
    if self._submission == nil then
      self._submission = EntityMod.new(self, nil)
    end
    return self._submission
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:UrlFetch():list() / client:UrlFetch():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function RsqSDK:UrlFetch(data)
  local EntityMod = require("entity.url_fetch_entity")
  if data == nil then
    if self._url_fetch == nil then
      self._url_fetch = EntityMod.new(self, nil)
    end
    return self._url_fetch
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Year():list() / client:Year():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function RsqSDK:Year(data)
  local EntityMod = require("entity.year_entity")
  if data == nil then
    if self._year == nil then
      self._year = EntityMod.new(self, nil)
    end
    return self._year
  end
  return EntityMod.new(self, data)
end




function RsqSDK.test(testopts, sdkopts)
  sdkopts = sdkopts or {}
  sdkopts = vs.clone(sdkopts)
  if type(sdkopts) ~= "table" then
    sdkopts = {}
  end

  testopts = testopts or {}
  testopts = vs.clone(testopts)
  if type(testopts) ~= "table" then
    testopts = {}
  end
  testopts["active"] = true

  vs.setpath(sdkopts, "feature.test", testopts)

  local sdk = RsqSDK.new(sdkopts)
  sdk.mode = "test"

  return sdk
end


return RsqSDK
